CI and build updates

ebiggers opened the pull request:
Bring in some of the improvements we've made in the hctr2 repository that are also applicable to the adiantum one, including:

* Get the benchmark tool building with the latest versions of the Android NDK and Meson
* Add some basic CI with GitHub Actions
* Improve the benchmark README file
* Avoid various warnings

ciphergoth commented on benchmark/src/util.h:
```diff
@@ -146,32 +150,32 @@ static inline void put_unaligned_be64(u64 v, void *p)
 static inline u16 rol16(u16 word, unsigned int shift)
 {
-	return (word << shift) | (word >> (16 - shift));
+	return (word << shift) | (word >> (-shift & 15));
```
What's this change for? Surprised you mask only the second shift in each case.

ebiggers replied:
As per the commit message for this change, this is specifically to avoid undefined behavior when rotating by 0 bits.  I have expanded the explanation in the commit message.

ciphergoth commented on benchmark/src/xts_benchmark_template.h:
```diff
@@ -8,12 +8,12 @@
 	const size_t bufsize = g_params.bufsize;
-	u8 *orig = malloc(bufsize);
-	u8 *ctext = malloc(bufsize);
+	u8 *orig = calloc(1, bufsize);
```
Why calloc not malloc?

ebiggers replied:
This is to avoid a compiler warning.  I've expanded the explanation in the commit message for this change.

ebiggers committed:
meson.build: add -fwrapv to compiler flags

The Linux kernel uses the -fwrapv compiler flag to make some undefined
behavior defined.  Without this flag, UBSAN reports an error in
third_party/linux-kernel/aes_ti.c.  We are generally trying to match the
Linux kernel settings, so add this flag.

ebiggers committed:
meson.build: use built-in optimization and warning options

Address the following meson warnings:

meson.build:20: WARNING: Consider using the built-in optimization level instead of using "-O2".
meson.build:20: WARNING: Consider using the built-in warning_level option instead of using "-Wall".

ebiggers committed:
meson.build: fix the custom_target commands

The "capture" parameter to custom_target is supposed to be a bool, not a
string.  The latest version of meson validates this strictly and fails
the build if it is a string.

ebiggers committed:
benchmark: avoid undefined behavior in rotate macros

Without this, rotations by 0 bits (such as those that can be done by the
RC5 code) trigger undefined behavior, and thus UBSAN warnings.

Note: it is not necessary to mask the first shift, since it is not
necessary to support rotation amounts that are too large.  The supported
rotation amounts are 0 through 'width - 1'.

ebiggers committed:
benchmark: fix a -Wunused-function warning

chacha_advance() isn't used in some configurations, and clang warns
about unused inline functions in .c files (unlike gcc), so clang warns:

    ../../src/chacha.c:159:20: warning: unused function 'chacha_advance' [-Wunused-function]

It's nontrivial to #ifdef this function out correctly, so just give it
the unused function attribute instead.

ebiggers committed:
benchmark: fix a -Wmaybe-uninitialized warning

gcc 11 warns:

../../src/xts_benchmark_template.h:76:16: warning: ‘ctext’ may be used uninitialized [-Wmaybe-uninitialized]
   76 |         ASSERT(memcmp(orig, ctext, bufsize));
../../src/xts_benchmark_template.h:76:9: note: in expansion of macro ‘ASSERT’
   76 |         ASSERT(memcmp(orig, ctext, bufsize));
      |         ^~~~~~

This is a false positive, but just zero-initialize the memory anyway.

ebiggers committed:
benchmark: add qemu-arm and qemu-aarch64 build types

These are helpful for testing.

ebiggers committed:
benchmark/README.md: simplify Preliminaries section

Give one command that installs all the packages, and assume that Python
3 is version 3.6 or later (it's been long enough).

ebiggers committed:
benchmark: bring in Android build fixes from hctr2 repository

Get the Android build working with the latest NDK, and update the README
file.

ebiggers committed:
Add ci.yml for GitHub Actions
